### PR TITLE
Feat/nft freemint

### PR DESCRIPTION
**Pull request by emarai**

### CODE REVIEW

File: contracts/shiden34/lib.rs [hunk 1]
```diff
@@ -77,7 +66,6 @@ pub mod shiden34 {
             name: String,
             symbol: String,
             base_uri: String,
-            max_supply: u64,
             price_per_mint: Balance,
```
Comment (Maar-io) [inline contracts/shiden34/lib.rs]: perhaps price is not needed since it is free mint

File: contracts/shiden34/lib.rs [hunk 2]
```diff
@@ -86,12 +74,25 @@ pub mod shiden34 {
...
             instance.payable_mint.price_per_mint = price_per_mint;
             instance.payable_mint.last_token_id = 0;
             instance.payable_mint.max_amount = 1;
+            instance.payable_mint.mint_end = false;
             instance
         }
+
+        #[ink(message)]
+        #[modifiers(only_owner)]
+        pub fn set_code(&mut self, code_hash: [u8; 32]) -> Result<(), PSP34Error> {
```
Comment (Maar-io) [inline contracts/shiden34/lib.rs:92]: why do you need this function?

File: contracts/shiden34/lib.rs [hunk 3]
```diff
@@ -176,15 +169,15 @@ pub mod shiden34 {
             set_sender(accounts.bob);
 
             assert_eq!(sh34.total_supply(), 0);
-            test::set_value_transferred::<ink::env::DefaultEnvironment>(PRICE);
+            test::set_value_transferred::<ink_env::DefaultEnvironment>(PRICE);
```
Comment (Maar-io) [inline contracts/shiden34/lib.rs]: ink_env is deprecated. Please use ink::env

File: logics/impls/payable_mint/payable_mint.rs [hunk 4]
```diff
@@ -71,39 +58,82 @@ where
...
+            self.data::<Data>().last_token_id += 1;
         }
 
+        self.data::<Data>().account_minted.insert(caller, &true);
         Ok(())
     }
 
     /// Mint next available token for the caller
     default fn mint_next(&mut self) -> Result<(), PSP34Error> {
+        self.check_amount(1)?;
```
Comment (Maar-io) [inline logics/impls/payable_mint/payable_mint.rs]: no need for this check. The `amount` is not parameter here

File: logics/impls/payable_mint/payable_mint.rs [hunk 5]
```diff
@@ -149,13 +178,14 @@ where
             String::from("baseUri"),
         );
         let mut token_uri = PreludeString::from_utf8(value.unwrap()).unwrap();
-        token_uri = token_uri + &token_id.to_string() + &PreludeString::from(".json");
+        token_uri = token_uri + "1" + &PreludeString::from(".json");
```
Comment (Maar-io) [inline logics/impls/payable_mint/payable_mint.rs]:
```suggestion
        token_uri = token_uri + &PreludeString::from("1.json");
```

File: logics/impls/payable_mint/payable_mint.rs [hunk 6]
```diff
@@ -182,34 +228,28 @@ where
     ) -> Result<(), PSP34Error> {
         if let Some(value) = (mint_amount as u128).checked_mul(self.data::<Data>().price_per_mint) {
             if transferred_value == value {
-                return Ok(())
+                return Ok(());
```
Comment (Maar-io) [inline logics/impls/payable_mint/payable_mint.rs]: not needed but it does not harm. Try `cargo clippy`

File: logics/impls/payable_mint/types.rs [hunk 7]
```diff
@@ -12,6 +12,8 @@ pub struct Data {
     pub max_supply: u64,
     pub price_per_mint: Balance,
```
Comment (Maar-io) [inline logics/impls/payable_mint/types.rs]: price not needed

File: logics/impls/payable_mint/payable_mint.rs [hunk 8]
```diff
@@ -167,6 +197,22 @@ where
...
+            .account_minted
+            .get(account_id)
+            .unwrap_or(false)
+    }
+
+    default fn get_mint_end(&self) -> bool {
+        self.data::<Data>().mint_end
+    }
+
+    default fn set_mint_end(&mut self, status: bool) -> Result<(), PSP34Error> {
```
Comment (Maar-io) [inline logics/impls/payable_mint/payable_mint.rs:213]:
```suggestion
    #[modifiers(only_owner)]
    default fn set_mint_end(&mut self, status: bool) -> Result<(), PSP34Error> {
```

File: logics/traits/payable_mint.rs [hunk 9]
```diff
@@ -2,10 +2,8 @@ use ink::prelude::string::String as PreludeString;
 
 use openbrush::{
     contracts::psp34::PSP34Error,
-    traits::{
-        AccountId,
-        Balance,
-    },
+    modifiers,
```
Comment (Maar-io) [inline logics/traits/payable_mint.rs]:
```suggestion
```

File: logics/traits/payable_mint.rs [hunk 10]
```diff
@@ -48,4 +46,14 @@ pub trait PayableMint {
     /// Get max number of tokens which could be minted per call
     #[ink(message)]
     fn get_max_mint_amount(&mut self) -> u64;
+
+    #[ink(message)]
+    #[modifiers(only_owner)]
```
Comment (Maar-io) [inline logics/traits/payable_mint.rs]:
```suggestion

```

File: logics/impls/payable_mint/payable_mint.rs [hunk 11]
```diff
@@ -71,38 +65,78 @@ where
     /// Mint one or more tokens
     #[modifiers(non_reentrant)]
     default fn mint(&mut self, to: AccountId, mint_amount: u64) -> Result<(), PSP34Error> {
+        assert_eq!(mint_amount, 1);
```
Comment (Maar-io) [inline logics/impls/payable_mint/payable_mint.rs]: this check is not needed and assert will make it hard to debug.
Check for the amount is done in the line below and amount is fixed to 1 in the `new()` call

File: contracts/shiden34/lib.rs (hunk 2 above)
Comment (emarai) [reply contracts/shiden34/lib.rs:92]: Our plan is to make the contract upgradable to fix production bugs until the contract is properly audited

**Pull request merged**